graph.go

```diff
@@ -95,7 +95,9 @@ type NodeRemover interface {
 // EdgeSetter is an interface for adding edges to a graph.
 type EdgeSetter interface {
 	// SetEdge adds an edge from one node to another.
-	// If the nodes do not exist, they are added.
+	// If the graph supports node addition the nodes
+	// will be added if they do not exist, otherwise
```

Comment:
This is a Go convention that centres around the fact that the type system is not strong enough to fully specify behavioural restrictions. The use of "will" is based on the fact that the docs are written for the user of the type more than the implementor (the user will read more). Because the docs say something will happen, the implementor then must implement that described behaviour.

The cnsequence of all this is that we do know what will be done.

Of course, implementors are free to not follow the exact specification, but then the behaviour of the routines we provide may not be as they are documented. Since we have decided that our graphs are (currently) simple, adding a self loop is illegal via our API (I do this privately in the community optimisation code, but comment that and don't use the graph API to do the addition).